[TVM][RUNTIME] A minimum example to generate external library wrappers for DSOModule

zhiics opened the pull request:
A minimum runtime for external codegen as mentioned in #4258

tqchen commented:
To followup on the discussion thread. While we understand that the current PR achieves the purpose of supporting external library by a different kind of wrapping, it does brings additional code that is largely duplicates with DSOModule.

Unless there is a strong reason to do so, I think we should not introduce the new Extern runtime. Instead, let us document the DSOModule's calling convention clearly and rewrite compilers to generate functions that can be loaded by the DSO module. That is, instead of generating a function with the extern calling convention, generate a function that takes the following signature

```c++
// the original foo you intended to generate
void foo_(float* a, int N, float* b, int M, float* c) {
   bar(...);
   foobar(...);
}

// DSOModule compatible c function that can be compiled by gcc.
extern "C" int foo(TVMValue* value, int *type_code, int nargs) {
   CHECK_EQ(nargs, 2);
   DLTensor* arg0 = static_cast<DLTensor*>(value.v_handle);
   DLTensor* arg1 = static_cast<DLTensor*>(value.v_handle);

   foo_(static_cast<float*>(args0->data), args0->shape[0],
           static_cast<float*>(args1->data), args1->shape[0]));
}
```

The main rationale behind the suggestion is that we want to avoid code specializations and reduce future technical debts. As we can see the current PR introduces special changes to several locations to support the new mode. Which increases the cost of maintaining the separate loading mechanism.

tqchen commented:
As a followup thread about what we might be able to benefit from in terms of external module, it would be great if we have another function that generates something like a virtual inst graph, which can be interpreted and calls into the library functions(just a bit like our graph runtime), see also https://discuss.tvm.ai/t/standardize-graphruntime-exports-into-a-single-dll/4667 

This kind of external module is something closer to support modules like tensorrt, or tf and would be a nice example to have.

zhiics commented:
@tqchen I removed everything, but just kept C APIs as a CSourceModule. I think this could be a minimal `runtime`/example that we can use for external library that needs to produce DSO. For the ones that need artifacts like Json (e.g TRT), let me think a little more.

comaniac commented:
@tqchen could you help review again? Here is a short summary of the current runtime after your suggestions:

* The runtime takes only one .so file including host and external kernels.
* The runtime is now unified with `DSOModule` so backend developers don't need to implement their own runtime (and `GetFunction`) anymore. Instead, they need to unpack `TVMValue` in the generated kernels.
* The [example](https://github.com/apache/incubator-tvm/pull/4280/files#diff-aab1542f9a9fec245df5ae78088f43ffR253) in this PR demonstrates a case that user backend will generate third-party library calls. This is applicable to runtime-less libraries such as MKL-DNN.
* For the external backend that has a runtime engine such as TRT, we can still codegen external functions like [this example](https://github.com/apache/incubator-tvm/pull/4280/files#diff-aab1542f9a9fec245df5ae78088f43ffR27), construct the subgraph in that function, and invoke the TRT engine to execute it on the fly.

We will refine the other codegen PR based on this one after merging.

tqchen commented:
Thanks @comaniac i feel that given it is already unified with the DSOModule, we don't have to call it a separate extern runtime, instead it is a way to interface external libraries into the current DSOModule runtime:)

It would still be great if we can add the other example that I mentioned, where we implement a customized runtime that have its own serialization mechanism(perhaps a graph like graph runtime) and the PackedFunc executes the corresponding functions by interpreting the serialized structure(e.g. graph). This kind of example will be closer to what people want, especially they want to serialize states that are other than sequence of API calls. @zhiics

zhiics commented:
yes, we don't need to call this as a runtime now. It actually just use DSOModule and CSourceModule. I think @comaniac was asking should we add this example (DSO style external library) first and then move to implement the customized runtime you are mentioning. Never mind, we can work on the customized runtime and update the PR. Thanks.

comaniac commented:
> It would still be great if we can add the other example that I mentioned, where we implement a customized runtime that have its own serialization mechanism(perhaps a graph like graph runtime) and the PackedFunc executes the corresponding functions by interpreting the serialized structure(e.g. graph). This kind of example will be closer to what people want, especially they want to serialize states that are other than sequence of API calls. @zhiics

I've added another example that mimics the graph-runtime-like programming model. This example assumes the external backend has its own runtime that accepts a graph representation such as JSON. Accordingly, the codegen would serialize the subgraph to a string like this [code snippet](https://github.com/apache/incubator-tvm/pull/4280/files#diff-aab1542f9a9fec245df5ae78088f43ffR128-R131) (I use a very simple representation instead of a complete JSON format in this example for simplification). Note that the codegen can either embed the graph to kernel code, or put the graph to a separate file and read it back in runtime.

comaniac commented:
cc @tqchen

yzhliu commented on tests/python/relay/test_external_runtime.py:
```diff
@@ -0,0 +1,492 @@
+    extern "C" void gcc_1_(float* gcc_input4, float* gcc_input5,
+                           float* gcc_input6, float* gcc_input7, float* out) {
+      float* buf_0 = (float*)malloc(4 * 100);
+      float* buf_1 = (float*)malloc(4 * 100);
```
shall we `free`?

zhiics replied:
Good point. Updated. Thanks.

yzhliu commented:
@tqchen would you take a look again?

tqchen commented:
@comaniac Thanks for the update! Can you move the ExampleJSONModule into a formal module that is compiled by the TVM runtime? Because normally, we will need to implement a SaveToBinary and LoadFromBinary instead of pumping things into a DSO file

comaniac commented:
> @comaniac Thanks for the update! Can you move the ExampleJSONModule into a formal module that is compiled by the TVM runtime? Because normally, we will need to implement a SaveToBinary and LoadFromBinary instead of pumping things into a DSO file

The current implementation expects external codegens to write the JSON string directly to the DSO module so that we can have a unified interface. Do you mean we should also provide the following JSON module other than DSO module for developers to use:

```c++
JSONModule : runtime::Module {
    void SaveToBinary();
    void LoadFromFile();
};
```

Now, user would have two JSON files if the user invoke json_module.SaveToBinary() and relay.build()? Would this mean that we have to change API of graph_runtime.create() to include the second JSON file?

Also, we will generate the following function in the .so.

```c++
// JSONModule compatible c function that can be compiled by gcc.
extern "C" int foo(TVMValue* value, int *type_code, int nargs, std::string json_file) {
   CHECK_EQ(nargs, 3);
   DLTensor* arg0 = static_cast<DLTensor*>(value.v_handle);
   DLTensor* arg1 = static_cast<DLTensor*>(value.v_handle);

   // Parse json_file and use "foo" as the key to get the subgraph JSON.
  std::string subgraph_json = ...;

   foo_(static_cast<float*>(args0->data), args0->shape[0],
           static_cast<float*>(args1->data), args1->shape[0]),
           subgraph_json);
}
```

So that the user codegen is expected to generate the following:

```c++
// the original foo you intended to generate
void foo_(float* a, int N, float* b, int M, float* c, std::string subgraph_json) {
    // Launch the external engine with subgraph_json and arguements.
}
```

tqchen commented:
The current interfacing example already looks great. For the JSONModule usecase, i was thinking about something like

```c++
JSONModule : runtime::Module {
    void SaveToBinary();
    void LoadFromFile();
    PackedFunc GetFunction(name) {
       if (name="tensortrun") {
         for each elem in json:
             if (elem.name == "func1") {
                run_trt_func1(func, elem.args);
             }  
            ...
      }
    }
};
```
In this case, because JSONModule directly interprets the json and calls into the runtime API, we no longer needs the gcc related components. This is an approach that will work better say if the serialized code is a TF blob plus some meta data and we just want to call into TF runtime.

tqchen commented:
Also note that my comment about ExampleJSONModule is mainly to serve as example ways to integrate for other external runtimes.

zhiics commented:
@tqchen I made a gtest for the simple example JSON runtime. Hopefully, this is good now. Please take a look when you have time.

zhiics commented on include/tvm/runtime/module.h:
```diff
@@ -111,7 +111,7 @@ class Module : public ObjectRef {
  * \endcode
  */
-class ModuleNode : public Object {
+class TVM_DLL ModuleNode : public Object {
```
@tqchen We added `TVM_DLL` so that the ExampleJsonModule can inherit it in the test. It looks that this is also used by other classes like `DeviceAPI` i.e.. Is this the correct change? otherwise, please advice.

tqchen replied:
sgtm

tqchen commented on tests/cpp/external_runtime_test.cc:
```diff
@@ -0,0 +1,245 @@
+    }
+  }
+
+  void run(int id, const std::vector<int>& inputs) {
```
Run, as per GoogleC style(camelcase)

tqchen commented:
Comments about json runtime. I would consider move the json runtime to 
```apps/extern_runtime``` to allow us to write examples even in python.  It would be great to see examples like:

```python
mod = DSOModule()
mod.import_module(JSONExampleModule())

mod.export_library("xyz.so")

mod = tvm.module.load("xyz.so")
# run that he code that can uses functions in the JSONExampleModule
```

comaniac commented:
@tqchen we have moved the example JSON runtime as well as the other two examplese to `apps/ext_runtime` as suggested. Please see [here](https://github.com/apache/incubator-tvm/pull/4280/files#diff-1a4d6732611c24f6793473a10b67b4a6R520-R526) for the use case of the example JSON runtime.

Please note that we need to change [get function call](https://github.com/apache/incubator-tvm/pull/4280/files#diff-5243f3909dfa484d85d3d8471f259169R399) in the graph runtime in order to search the implementation in the imported module. Please also help check if it's ok to do so.

Thanks.

u99127 commented:
> Comments about json runtime. I would consider move the json runtime to
> `apps/extern_runtime` to allow us to write examples even in python. It would be great to see examples like:
> 

> ```python
> mod = DSOModule()
> mod.import_module(JSONExampleModule())
> 
> mod.export_library("xyz.so")
> 
> mod = tvm.module.load("xyz.so")
> # run that he code that can uses functions in the JSONExampleModule
> ```
> 
> The main goal is to provide readers an easy reference point about how to add their own runtimes.

Isn't it also to provide a standardized structure for them to integrate their runtimes ? Are we expecting such integrations to happen via the 3rd party directory mechanism which is currently really for build dependencies. Giving an example integration shows how it can be done, providing with an initial location and standard mechanism for the simple toy integration. 

In my opinion a simple integration should provide a template to follow. Putting this in examples appears to suggest that folks can do what they want with the 3rd party runtime integration.

> 
> As a followup, we can add docs/dev/custom_codegen_runtime.rst that introduces both the compile to DSOModule method and the extern json runtime example, with reference to the folder.

I'm puzzled by this direction. 

What is missing for me in this pull request is actually a directory structure or something that indicates where and how custom run times can go in. This suggests that we could have a full custom run time sitting in the app / examples folder  ? Instead shouldn't a custom run time really be in the 3rd party folder rather than sitting in the app folder or indeed under tvm/runtime/3rdparty/ or some such.


Ramana

tqchen commented on apps/ext_runtime/src/ext_json_rt.cc:
```diff
@@ -0,0 +1,264 @@
+#include <tvm/runtime/memory.h>
+#include <tvm/runtime/module.h>
+#include <tvm/runtime/ndarray.h>
+#include <tvm/runtime/object.h>
```
We will need more comments in this file, in particular, how can others implement there own customized runtime, list the functions to change, for example:

- GetFunction
- SaveToBinary
- LoadFromBinary

zhiics replied:
I added the comments for these three fields. We will put more detailed explanation and usage in the tutorial after the example is merged.

tqchen commented:
The directory convention should be something we document in the documents. Currently a convention location for the project would be under ```src/runtime/contrib```.

However, as long as the code are linked to the tvm runtime. The example can run. @u99127 are you suggesting that we should move the folder directly under ```src/runtime/contrib```?

zhiics commented:
yes. Currently, this is an example that provides a demo to tell people how we can write minimal DSO and JSON style runtime. I think it makes sense to put them under `apps` as then they will show up on the `tvm.ai` website. Or production ready codegen/runtime, we will put it under `src/runtime/contrib` as @tqchen mentioned above.

These examples are real code that can be compiled and executed. We will likely add DNNL later on when we update the other PR (#4258). And that one will be under `src/runtime/contrib`.

comaniac commented:
@tqchen thanks for the comments. Do you have any other suggestions to this PR, or could we merge it?

u99127 commented:
> The directory convention should be something we document in the documents. Currently a convention location for the project would be under `src/runtime/contrib`.
> 

And the example added should fit with the documentation ?

> However, as long as the code are linked to the tvm runtime. The example can run. @u99127 are you suggesting that we should move the folder directly under `src/runtime/contrib`?

Yes or the example should be arranged so that this is a canonical example for integration with apps/examples and src/runtime/contrib.

My understanding is that the apps/example that's being put together is for the end user to consume TVM or TVM integrated with the 3rd party library integration, but the intermediate step of providing the framework for the integrator is not being shown or is not obvious to me.

regards
Ramana

u99127 commented:
> yes. Currently, this is an example that provides a demo to tell people how we can write minimal DSO and JSON style runtime. I think it makes sense to put them under `apps` as then they will show up on the `tvm.ai` website. Or production ready codegen/runtime, we will put it under `src/runtime/contrib` as @tqchen mentioned above.
> 

My view is that having the example implementation structured in the way that the production library example would land is a good thing.


> These examples are real code that can be compiled and executed. We will likely add DNNL later on when we update the other PR (#4258). And that one will be under `src/runtime/contrib`.

It seems to me that an end-to-end simple example should be a simple reference for integration rather than starting with a specific integration example like DNNL and worrying about any specifics that maybe from an integration of DNNL.

regards
Ramana

tqchen commented:
OK, I think @u99127 has a good point. 

Let us do the following things:
- Move  apps/ext_runtime to ```src/runtime/contrib/example_ext_runtime```
- Add option USE_EXAMPLE_EXT_RUNTIME, which switchs off by default, but enabled during testcode
- Add tests to the original unittest folder.
- Add more comments in the json runtime code itself, to talk about what a new developer should do.

Sorry for going back and forth, but hopefully we can make the final structure better for our developers.

zhiics commented:
@u99127 Thanks for your comment.

@tqchen No worries. We actually had the same impression that the example should be put under contrib. But it looks that app dir also fine as this is really just an example. But anyways, we now moved it to `contrib` and added comments to the file. Please take another look.

tqchen commented on src/runtime/contrib/exmaple_ext_runtime/example_ext_runtime.cc:
```diff
@@ -0,0 +1,338 @@
+   *
+   * \return The created json module.
+   */
+  static Module LoadFromFile(const std::string& json,
```
The first argument of LoadFromFile is the file path. I feel that we can omit LoadFromFile(as we mainly export through dll, and LoadFromBinary is enough)

u99127 commented on CMakeLists.txt:
```diff
@@ -231,6 +231,12 @@ if(USE_VM_PROFILER)
 
+if(USE_EXAMPLE_EXT_RUNTIME)
+  message(STATUS "Build with Relay VM profiler support...")
+  file(GLOB RUNTIME_EXAMPLE_EXTERNAL_SRCS src/runtime/contrib/exmaple_ext_runtime/*.cc)
```
example ? Instead of exmaple ?

Message should say build with example external runtime and not relay vm profiler support .

u99127 commented:
> @u99127 Thanks for your comment.

Thanks @zhiics and @tqchen.

comaniac commented:
Comment fixed.
@tqchen somehow the CI doesn't update the config.cmake to enable the example runtime. Would you take a look when you get a chance? Thanks.

tqchen commented:
@comaniac please send another PR first with the updated Jenkinsfile. CI will only respect Jenkinsfile that is already in master

u99127 commented:
There are still a couple of changes requested in the review that need to be addressed. It would be good to clean this up with a separate pull request.

tqchen commented:
@u99127 please https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

u99127 commented:
> 
> 
> @u99127 please https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

Sorry about the noise, it seems I misread the pull request this morning as being merged but now I realize it was because of the CI changes that got merged and not the actual PR.

zhiics commented:
@u99127 Thanks. Cloud you please explicitly approve if everything looks good to you?

u99127 commented:
LGTM - I don't see an actual "Approve" button if clicking that's expected from a workflow perspective.

Ramana

zhiics commented:
@u99127 I see, thank you. You can find the "button" like this:

https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/approving-a-pull-request-with-required-reviews

comaniac commented:
@tqchen would you mind to take a final look and merge if it's all good?
Thanks!

tqchen commented:
Thanks @comaniac @zhiics @u99127 !

tqchen commented:
@zhiics @comaniac please followup to fix the windows build error here https://dev.azure.com/tvmai/tvm/_build/results?buildId=3417
